pkg/datapath/loader/loader.go

```diff
@@ -99,10 +109,150 @@ func removeEndpointRoute(ep datapath.Endpoint, ip net.IPNet) error {
+		}
+		if option.Config.EnableIPv6 {
+			directRoutingIPv6 := nodePortIPv6Addrs[directRoutingIface]
+			opts["IPV6_DIRECT_ROUTING_1"] = sliceToBe32(directRoutingIPv6[0:4])
+			opts["IPV6_DIRECT_ROUTING_2"] = sliceToBe32(directRoutingIPv6[4:8])
+			opts["IPV6_DIRECT_ROUTING_3"] = sliceToBe32(directRoutingIPv6[8:12])
+			opts["IPV6_DIRECT_ROUTING_4"] = sliceToBe32(directRoutingIPv6[12:16])
+			nodePortIPv6 := nodePortIPv6Addrs[ifName]
+			opts["IPV6_NODEPORT_1"] = sliceToBe32(nodePortIPv6[0:4])
+			opts["IPV6_NODEPORT_2"] = sliceToBe32(nodePortIPv6[4:8])
+			opts["IPV6_NODEPORT_3"] = sliceToBe32(nodePortIPv6[8:12])
+			opts["IPV6_NODEPORT_4"] = sliceToBe32(nodePortIPv6[12:16])
+		}
+	}
+
+	// Among string substitutions, only the calls map name is specific to each
+	// attachment interface.
+	strings = nullifyStringSubstitutions(strings)
+	callsMapHostDevice := bpf.LocalMapName(callsmap.HostMapName, uint16(ep.GetID()))
+	strings[callsMapHostDevice] = bpf.LocalMapName(callsmap.NetdevMapName, uint16(ifIndex))
+
+	return hostObj.Write(dstPath, opts, strings)
+}
+
+func (l *Loader) reloadHostDatapath(ctx context.Context, ep datapath.Endpoint, objPath string) error {
+	symbols := []string{symbolToHostEp, symbolFromHostEp}
+	directions := []string{dirIngress, dirEgress}
+	objPaths := []string{objPath, objPath}
+
+	nodePortIPv4Addrs := node.GetNodePortIPv4AddrsWithDevices()
+	nodePortIPv6Addrs := node.GetNodePortIPv6AddrsWithDevices()
+
+	interfaceNames := make([]string, 2)
+	interfaceNames[0] = ep.InterfaceName()
+	interfaceNames[1] = interfaceNames[0]
+	for _, device := range option.Config.Devices {
+		netdevObjPath := path.Join(ep.StateDir(), hostEndpointNetdevPrefix+device+".o")
+		if err := patchHostNetdevDatapath(ep, objPath, netdevObjPath, device, nodePortIPv4Addrs, nodePortIPv6Addrs); err != nil {
+			return err
+		}
+		objPaths = append(objPaths, netdevObjPath)
+
+		interfaceNames = append(interfaceNames, device)
+		symbols = append(symbols, symbolFromHostNetdevEp)
+		directions = append(directions, dirIngress)
+		if option.Config.EnableNodePort {
+			interfaceNames = append(interfaceNames, device)
+			symbols = append(symbols, symbolToHostNetdevEp)
+			directions = append(directions, dirEgress)
+			objPaths = append(objPaths, netdevObjPath)
+		}
+
+	}
+
+	for i, interfaceName := range interfaceNames {
+		symbol := symbols[i]
+		if err := l.replaceDatapath(ctx, interfaceName, objPaths[i], symbol, directions[i]); err != nil {
+			scopedLog := ep.Logger(Subsystem).WithFields(logrus.Fields{
+				logfields.Path: objPath,
+				logfields.Veth: interfaceName,
+			})
+			// Don't log an error here if the context was canceled or timed out;
+			// this log message should only represent failures with respect to
+			// loading the program.
+			if ctx.Err() == nil {
+				scopedLog.WithError(err).Warningf("JoinEP: Failed to load program for host endpoint (%s)", symbol)
+			}
+			return err
```

Comment:
nit: Did you consider the potential implications of returning on the first error here?

Not sure the full set of cases this might affect, but here's a straw man:

If a user specifies a bad device name in the list of devices, then it looks like this function would successfully load the BPF programs for the base host functionality, as well as any "good" device names in the list prior to the "bad" device name, then fail to load the prog onto the bad device name and any subsequent devices. Though the error would be propagated up as well, not sure how that ends up impacting the higher-layer logic.

(I don't *think* we have a lot to worry about here, more just seeing if you'd considered it)

Reply:
I will add a sanity check on the attach links (just checking they exist). We already have a similar check before calling `init.sh`.

Nevertheless, our current code would also fail in a similar manner if there's an attach error or if the link disappears between the sanity check and the actual loading (since `init.sh` has `set -e`). I'd rather not fix/change that behavior in this PR, but maybe we should skip any failing device in a follow up PR instead of skipping all devices on any fail.